docs/guide/dependency-graph.md

```diff
@@ -0,0 +1,96 @@
+engine would need to add `n*(n+1)/2 ≈ n²` arcs in the graph. This
+value grows much faster than the size of data, meaning the engine
+would not be able to handle large data sets efficiently.
+
+A solution to this problem comes from the observation that there is
+a way to rewrite the above formulas to equivalent ones, which will
+be more compact to represent. Specifically, the following formulas
+would compute the same values as the ones provided previously:
+
+* `B1=A1`
+* `B2=B1+A2`
+* `B3=B2+A3`
+* ...
+* `B100=B99+A100`
+
+Whereas this example is too specialized to provide a useful rule
+for optimization, it shows the main idea behind efficient handling
+of multiple ranges: **to represent a range as a composition of
+smaller ranges.**
+
+In the adopted implementation, every time the engine encounters a
+range, say `B5:D20`, it checks if it has already considered the
+range which is one row shorter. In this example, it would be `B5:D19`.
+If so, then it represents `B5:D20` as the composition of a range
+`B5:D19` and three cells in the last row: `B20`,`C20` and `D20`.
+
+<img :src="$withBase('/ranges.png')">
+
+More generally, the result of any associative operation is obtained
+as the result of operations for these small rows. There are many
+examples of such associative functions: `SUM`, `MAX`, `COUNT`, etc.
+As one range can be used in different formulas, we can reuse its
+node and avoid duplicating the work during computation.
+
+## Getting the graph neighbors of a cell
+
+HyperFormula API includes methods that reveal parts of the dependency graph. In particular, they return the graph neighbors of a given cell.
+- [getCellPrecedents](../api/classes/hyperformula.html#getcellprecedents) for getting the in-neighbors
+- [getCellDependents](../api/classes/hyperformula.html#getcelldependents) for getting the out-neighbors
+
+## Getting all precedents or dependents of a cell
+
+Some applications may require processing all the precedents or dependents of a specified cell (not only immediate). In terms of graph theory, this problem can be stated as finding all the nodes reachable from the source node.
+
+Methods `getCellPrecedents` and `getCellDependents` return only the immediate precedents and dependents of a cell respectively, but they can be used to implement the algorithm for the reachability problem applying the [BFS](https://en.wikipedia.org/wiki/Breadth-first_search) -like approach:
+
+```
+ 1      AllCellPrecedents={start}
+ 2      let Q be an empty queue
+ 4      Q.enqueue(start)
+ 5      while Q is not empty do
+ 6          cell := Q.dequeue()
+ 7          S := getCellPrecedents(cell)
+ 9          for all cells c in S do:
+10              if c is not in AllCellPrecedents then:
+11                  insert w to AllCellPrecedents
+12                  Q.enqueue(c)
+```
+
+`AllCellDependents` algorithm can be obtained by changing `getCellPrecedents` usage to `getCellDependents` in the pseudocode above.
```

Comment:
```suggestion
## Getting the immediate precedents of a cell or a range

To get the immediate precedents of a cell or a range (the in-neighbors of the cell node or the range node), use the [`getCellPrecedents()`](../api/classes/hyperformula.html#getcellprecedents) method:

\```
const hfInstance = HyperFormula.buildFromArray( [ ['1', '=A1', '=A1+B1'] ] );

hfInstance.getCellPrecedents({ sheet: 0, col: 2, row: 0});
// returns [{ sheet: 0, col: 0, row: 0}, { sheet: 0, col: 1, row: 0}]
\```

## Getting the immediate dependents of a cell or a range

To get the immediate dependents of a cell or a range (the out-neighbors of the cell node or the range node), use the [`getCellDependents()`](../api/classes/hyperformula.html#getcelldependents) method:

\```
const hfInstance = HyperFormula.buildFromArray( [ ['1', '=A1', '=A1+B1'] ] );

hfInstance.getCellDependents({ sheet: 0, col: 2, row: 0});
// returns [{ sheet: 0, col: 0, row: 0}, { sheet: 0, col: 1, row: 0}]
\```

## Getting all precedents of a cell or a range

To get all precedents of a cell or a range (all precedent nodes reachable from the cell node or the range node), use the [`getCellPrecedents()`](../api/classes/hyperformula.html#getcellprecedents) method to implement a [Breadth-first search (BFS)](https://en.wikipedia.org/wiki/Breadth-first_search) algorithm:

\```
 1      AllCellPrecedents={start}
 2      let Q be an empty queue
 4      Q.enqueue(start)
 5      while Q is not empty do
 6          cell := Q.dequeue()
 7          S := getCellPrecedents(cell)
 9          for all cells c in S do:
10              if c is not in AllCellPrecedents then:
11                  insert w to AllCellPrecedents
12                  Q.enqueue(c)
\```

## Getting all dependents of a cell or a range

To get all dependents of a cell or a range (all dependent nodes reachable from the cell node or the range node), use the [`getCellDependents()`](../api/classes/hyperformula.html#getcelldependents) method to implement a [Breadth-first search (BFS)](https://en.wikipedia.org/wiki/Breadth-first_search) algorithm:

\```
 1      AllCellDependents={start}
 2      let Q be an empty queue
 4      Q.enqueue(start)
 5      while Q is not empty do
 6          cell := Q.dequeue()
 7          S := getCellDependents(cell)
 9          for all cells c in S do:
10              if c is not in AllCellDependents then:
11                  insert w to AllCellDependents
12                  Q.enqueue(c)
\```
```

Reply:
If we think it's an actual practical problem that the user can come across ("How do I get all dependents of a range if there's no dedicated method?"), I'd make each action into an independent procedure

Reply:
It as an actual problem. We got questions from 3 different users about getting precedents/dependents of a cell.